doc/content/toolstack/features/NUMA/index.md

```diff
@@ -62,39 +62,59 @@ The Xen scheduler supports 2 kinds of constraints:
 * hard pinning: a vCPU may only run on the specified set of pCPUs and nowhere else
 * soft pinning: a vCPU is *preferably* run on the specified set of pCPUs, but if they are all busy then it may run elsewhere
 
-The former is useful if you want strict separation, but it can potentially leave part of the system idle while another part is bottlenecked with lots of vCPUs all competing for the same limited set of pCPUs.
+Hard pinning can be used to partition the system. But, it can potentially leave part of the system idle while another part is bottlenecked by many vCPUs competing for the same limited set of pCPUs.
 
-Xen does not migrate workloads between NUMA nodes on its own (the Linux kernel does), although it is possible to achieve a similar effect with explicit migration.
-However migration introduces additional delays and is best avoided for entire VMs.
+Xen does not migrate workloads between NUMA nodes on its own (the Linux kernel can). Although, it is possible to achieve a similar effect with explicit migration.
+However, migration introduces additional delays and is best avoided for entire VMs.
 
-The latter (soft pinning) is preferred: running a workload now, even on a potentially suboptimal pCPU (higher NUMA latency) is still better than not running it at all and waiting until a pCPU is freed up.
+Therefore, soft pinning is preferred: Running on a potentially suboptimal pCPU that uses remote memory could still be better than not running it at all until a pCPU is free to run it.
 
-Xen will also allocate memory for the VM according to the vCPU (soft) pinning: if the vCPUs are pinned only to NUMA nodes A and B, then it will allocate the VM's memory from NUMA nodes A and B (in a round-robin way, resulting in interleaving).
+Xen will also allocate memory for the VM according to the vCPU (soft) pinning: If the vCPUs are pinned to NUMA nodes A and B, Xen allocates memory from NUMA nodes A and B in a round-robin way, resulting in interleaving.
```

Comment:
If Xen does this, why are parallel memory allocations an issue? This should currently allows xenopsd to control what NUMA nodes Xen reserves memory for the domain, as xenopsd decides the vcpu pinning before the reservation happens while holding a lock, which serializes the operation.